File: Java/chapter_14/app/src/main/java/com/glella/criminalintent/CrimeLab.java
package com.glella.criminalintent;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.glella.criminalintent.database.CrimeBaseHelper;
import com.glella.criminalintent.database.CrimeCursorWrapper;
import com.glella.criminalintent.database.CrimeDbSchema.Crimetable;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class CrimeLab {
    private static CrimeLab sCrimeLab;
    //private List<Crime> mCrimes;
    private Context mContext;
    private SQLiteDatabase mDatabase;

    public static CrimeLab get(Context context) {
        if (sCrimeLab == null) {
            sCrimeLab = new CrimeLab(context);
        }
        return sCrimeLab;
    }

    private CrimeLab(Context context) {
        mContext = context.getApplicationContext();
        mDatabase = new CrimeBaseHelper(mContext).getWritableDatabase();
        //mCrimes = new ArrayList<Crime>();
    }

    public List<Crime> getCrimes() {
        List<Crime> crimes = new ArrayList<>();

        CrimeCursorWrapper cursor = queryCrimes(null, null);

        try {
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                crimes.add(cursor.getCrime());
                cursor.moveToNext();
            }
        } finally {
            cursor.close();
        }
        return crimes;
    }

    public Crime getCrime(UUID id) {
        CrimeCursorWrapper cursor = queryCrimes(Crimetable.Cols.UUID + " = ?", new String[]{id.toString()});

        try {
            if (cursor.getCount() == 0) {
                return null;
            }

            cursor.moveToFirst();
            return cursor.getCrime();
        } finally {
            cursor.close();
        }
    }

    public void addCrime(Crime c) {
        //mCrimes.add(c); //old line

        ContentValues values = getContentValues(c);
        mDatabase.insert(Crimetable.NAME, null, values);
    }

    private static ContentValues getContentValues(Crime crime) {
        ContentValues values = new ContentValues();
        values.put(Crimetable.Cols.UUID, crime.getId().toString());
        values.put(Crimetable.Cols.TITLE, crime.getTitle());
        values.put(Crimetable.Cols.DATE, crime.getDate().getTime());
        values.put(Crimetable.Cols.SOLVED, crime.isSolved() ? 1 : 0);

        return values;
    }

    public void updateCrime(Crime crime) {
        String uuidString = crime.getId().toString();
        ContentValues values = getContentValues(crime);

        mDatabase.update(Crimetable.NAME, values, Crimetable.Cols.UUID + " = ?", new String[] {uuidString});
    }

    private CrimeCursorWrapper queryCrimes(String whereClause, String[] whereArgs) {
        Cursor cursor = mDatabase.query(
                Crimetable.NAME,
                null, // columns - null selects all columns
                whereClause,
                whereArgs,
                null, // groupBy
                null, // having
                null // orderBy
        );
        return new CrimeCursorWrapper(cursor);
    }

}

File: Kotlin/chapter_15_challenge/app/src/main/java/com/glella/criminalintent/CrimeLab.kt
package com.glella.criminalintent


import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import com.glella.criminalintent.database.CrimeBaseHelper
import com.glella.criminalintent.database.CrimeCursorWrapper
import com.glella.criminalintent.database.CrimeDbSchema
import java.util.*

object CrimeLab {
    var mCrimes = ArrayList<Crime>()
    var db: SQLiteDatabase

    val tablename = CrimeDbSchema.Crimetable.NAME
    val uuidColumn = CrimeDbSchema.Crimetable.Cols.UUID
    val titleColumn = CrimeDbSchema.Crimetable.Cols.TITLE
    val dateColumn = CrimeDbSchema.Crimetable.Cols.DATE
    val solvedColumn = CrimeDbSchema.Crimetable.Cols.SOLVED
    val suspectColumn = CrimeDbSchema.Crimetable.Cols.SUSPECT
    val suspectIDColumn = CrimeDbSchema.Crimetable.Cols.SUSPECTID
    val suspectPhoneColumn = CrimeDbSchema.Crimetable.Cols.SUSPECTPHONE

    init {
        val context: Context = ApplicationContextProvider.applicationContext()
        db = CrimeBaseHelper(context).writableDatabase
        mCrimes = getCrimes() as ArrayList<Crime>

    }

    fun getCrimes(): List<Crime> {
        val crimes = ArrayList<Crime>()

        val cursor = queryCrimes(null, null)

        try {
            cursor.moveToFirst()
            while (!cursor.isAfterLast()) {
                crimes.add(cursor.getCrime())
                cursor.moveToNext()
            }
        } finally {
            cursor.close()
        }
        return crimes
    }


    fun getCrime(id: UUID): Crime? {
        val cursor = queryCrimes("$uuidColumn = ?", arrayOf(id.toString()))

        try {
            if (cursor.count == 0) {
                return null
            }

            cursor.moveToFirst()
            return cursor.getCrime()
        } finally {
            cursor.close()
        }
    }

    fun addCrime(c: Crime) {
        //mCrimes.add(c); //old line

        val values = getContentValues(c)
        db.insert(tablename, null, values)
    }

    fun updateCrime(crime: Crime) {
        val uuidString = crime.mId.toString()
        val values = getContentValues(crime)

        db.update(tablename, values, uuidColumn + " = ?", arrayOf<String>(uuidString))
    }

    private fun getContentValues(crime: Crime): ContentValues {
        val values = ContentValues()
        values.put(uuidColumn, crime.mId.toString())
        values.put(titleColumn, crime.mTitle)
        values.put(dateColumn, crime.mDate!!.getTime())
        values.put(solvedColumn, if (crime.mSolved) 1 else 0)
        values.put(suspectColumn, crime.mSuspect)
        values.put(suspectIDColumn, crime.mSuspectID)
        values.put(suspectPhoneColumn, crime.mSuspectPhone)

        return values
    }

    private fun queryCrimes(whereClause: String?, whereArgs: Array<String>?): CrimeCursorWrapper {
        val cursor = db.query(
                tablename, // having
                null // orderBy
                , // columns - null selects all columns
                whereClause,
                whereArgs, null, null, null
        )// groupBy
        return CrimeCursorWrapper(cursor)
    }

}